pkg/apis/core/types.go

```diff
@@ -467,6 +467,16 @@ type PersistentVolumeSpec struct {
 	// This is an alpha feature and may change in the future.
 	// +optional
 	VolumeMode *PersistentVolumeMode
+	// NodeAffinity defines constraints that limit what nodes this volume can be accessed from.
+	// This field influences the scheduling of pods that use this volume.
+	// +optional
+	NodeAffinity *VolumeNodeAffinity
+}
+
+// VolumeNodeAffinity defines constraints that limit what nodes this volume can be accessed from.
+type VolumeNodeAffinity struct {
+	// Required specifies hard node constraints that must be met.
+	Required *NodeSelector
```

Comment:
I did not follow all the discussion closely. Why it changed from `RequiredDuringSchedulingIgnoredDuringExecution` to simple `Required`? What happens when admin edits the selector when a pod is already scheduled / running? Should it be allowed?

Reply:
I think the "RequiredDuringExecution" clause doesn't make sense for volumes because:
- That means the storage backend made the volume inaccessible on that node while it was still being used?  I don't think any sane backend would do that and cause data corruption for the application.
- Maybe the storage backend migrated the data to another node and but it's still globally accessible.  Then that becomes a preference instead of a requirement.  Also, I don't see a good reason why a backend would do that because it decrease workload performance.

In that case, maybe it is good to disallow modifying the Required field.  I can envision future scenarios for allowing changes to a Preferred field though.

Reply:
In general, disallow field updates until/unless you know exactly what they are for.

Reply:
Yup, it disallows updates now.  Hm but actually this will make it difficult for users to upgrade from alpha -> beta, or for automatically upgrading zonal PVs like GCE PDs to the beta format.

I think what I need to do is:
- Allow setting NodeAffinity if it was previously non-nil
- Disallow modifying NodeAffinity once its set.